// chromium-bidi/lib/cjs/bidiMapper/domains/context/browsingContextStorage.js
"use strict";
/**
 * Copyright 2022 Google LLC.
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowsingContextStorage = void 0;
const protocol_js_1 = require("../../../protocol/protocol.js");
class BrowsingContextStorage {
    static #contexts = new Map();
    static getTopLevelContexts() {
        return Array.from(BrowsingContextStorage.#contexts.values()).filter((c) => c.parentId === null);
    }
    static removeContext(contextId) {
        BrowsingContextStorage.#contexts.delete(contextId);
    }
    static addContext(context) {
        BrowsingContextStorage.#contexts.set(context.contextId, context);
        if (context.parentId !== null) {
            BrowsingContextStorage.getKnownContext(context.parentId).addChild(context);
        }
    }
    static hasKnownContext(contextId) {
        return BrowsingContextStorage.#contexts.has(contextId);
    }
    static findContext(contextId) {
        return BrowsingContextStorage.#contexts.get(contextId);
    }
    static getKnownContext(contextId) {
        const result = BrowsingContextStorage.findContext(contextId);
        if (result === undefined) {
            throw new protocol_js_1.Message.NoSuchFrameException(`Context ${contextId} not found`);
        }
        return result;
    }
}
exports.BrowsingContextStorage = BrowsingContextStorage;
