// chromium-bidi/lib/cjs/utils/log.js
"use strict";
/**
 * Copyright 2021 Google LLC.
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.log = exports.LogType = void 0;
var LogType;
(function (LogType) {
    LogType["system"] = "System";
    LogType["bidi"] = "BiDi Messages";
    LogType["browsingContexts"] = "Browsing Contexts";
    LogType["cdp"] = "CDP";
    LogType["commandParser"] = "Command parser";
})(LogType = exports.LogType || (exports.LogType = {}));
function log(logType) {
    return (...messages) => {
        console.log(logType, ...messages);
        // Add messages to the Mapper Tab Page, if exists.
        // Dynamic lookup to avoid circlular dependency.
        if ('MapperTabPage' in globalThis) {
            globalThis['MapperTabPage'].log(logType, ...messages);
        }
    };
}
exports.log = log;
